core-api/src/main/java/com/optimizely/ab/config/audience/UserAttribute.java

```diff
@@ -76,14 +81,34 @@ public Boolean evaluate(ProjectConfig config, Map<String, ?> attributes) {
         Object userAttributeValue = attributes.get(name);
 
         if (!"custom_attribute".equals(type)) {
-            MatchType.logger.error(String.format("condition type not equal to `custom_attribute` %s", type));
+            logger.error("Audience condition \"{}\" has an unknown condition type: {}", this.toString(), type);
             return null; // unknown type
         }
         // check user attribute value is equal
         try {
-            return MatchType.getMatchType(match, value).getMatcher().eval(userAttributeValue);
+            Match matchType = MatchType.getMatchType(match, value).getMatcher();
+            Boolean result = matchType.eval(userAttributeValue);
+
+            if (!(matchType instanceof NullMatch) && result == null) {
+                if (!attributes.containsKey(name)) {
+                    //Missing attribute value
+                    logger.debug("Audience condition \"{}\" evaluated as UNKNOWN because no user attribute for \"{}\" key is passed", this.toString(), name);
+                } else {
+                    //if attribute value is not valid
+                    logger.warn(
+                        "Audience condition \"{}\" evaluated as UNKNOWN because the value for user attribute \"{}\" is inapplicable: \"{}\"",
```

Comment:
We ended up taking y'all's suggestion of logging the name of the offending type! Can we go with
```
"Audience condition \"{}\" evaluated to UNKNOWN because a value of type \"{}\" was passed for user attribute \"{}\""
```
or if we were given a null attribute value:
```
"Audience condition \"{}\" evaluated to UNKNOWN because a null value was passed for the named user attribute"
```
It may be nontrivial to figure out what the type is. i think you may have been discussing with @loganlinn? Let me know if I can help or if we need to change tack.